enable fabric latency perf tests

SeanNijjar opened the pull request:
### Ticket
#20217 

### Problem description
Hop latency is an important performance metric for fabric. The latency tests we had were not mainlined yet. This PR finally mainlines them :) 

### What's changed
Added a simple latency microbenchmark. Here's a summary of some results (note: 1 hop = noc + eth latency):

- single flit hop latency: ~760 ns
- 4k payload hop latency: ~1300 ns
- fused payload + seminc packet single flit  ~770 ns

The above numbers are approximate given that worker placement hasn't been optimized yet for these workloads. 

There's decently large variation on the end-to-end latency (all-hops) of around 15%. 

### Future work 
- tighten up worker placement to isolate out noc distance latency effects
  - Also calculate away the noc latencies from the final numbers

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI: https://github.com/tenstorrent/tt-metal/actions/runs/14384690305
- [x] Microbench: https://github.com/tenstorrent/tt-metal/actions/runs/14385276089
- [x] BH PC: https://github.com/tenstorrent/tt-metal/actions/runs/14385400777
- [x] New/Existing tests provide coverage for changes

Closes #20217

SeanNijjar committed:
analysis script

WIP add fabric latency ubench

WIP chip skew test

fixes

fixed some bugs

further fixes - test working now

fixes and numbers

tighten bounds

cleanup

cleanup

SeanNijjar committed:
fixes and updates to numbers

SeanNijjar committed:
reorg

SeanNijjar committed:
update tests

SeanNijjar committed:
remove unnecessary file

SeanNijjar committed:
remove dprint

blozano-tt commented on .github/workflows/metal-run-microbenchmarks-impl.yaml:
```diff
@@ -55,7 +55,7 @@ jobs:
             runs-on: ["arch-wormhole_b0", "pipeline-perf", "config-t3000", "in-service"],
             name: "T3K ubench",
-            cmd: "pytest -svv tests/tt_metal/microbenchmarks/ethernet/test_fabric_edm_bandwidth.py",
+            cmd: "pytest -svv tests/tt_metal/microbenchmarks/ethernet/test_fabric_edm_bandwidth.py; pytest -svv tests/tt_metal/microbenchmarks/ethernet/test_1d_fabric_latency.py",
```
@williamlyTT will this cause any problems with the error tracking in superset?

I'm guessing no, but wanted to double check with you.

tt-rkim replied:
Could this also be a folder call?

tt-rkim replied:
That would ensure no weirdness w/ superset. Shouldn't be with William's new randomization logic, but still, @SeanNijjar why are they separate flies? Will there be more `tests/tt_metal/microbenchmarks/ethernet/` that aren't meant to be called here?

williamlyTT replied:
Nope there won't be any issues with error tracking; each pytest invocation will generate one xml test report and all reports get analyzed into superset.

SeanNijjar replied:
Can I please get an update on how I should specify the test if I need to change it for the PR? Merge is blocked on approval here. Thanks

daminakaTT replied:
I have no idea "superset" in this context, but why not having different job for this specific command? if the first test hang, the latter one is not executed

SeanNijjar replied:
good point @daminakaTT . @williamlyTT @tt-rkim is this possible with the current setup? I think it makes sense to split this into 2: "T3K Fabric BW Ubench" and "T3K Fabric Latency Ubench"

williamlyTT replied:
Yes you just need to add a new test group with the second pytest cmd

SeanNijjar replied:
Alright this thread is resolved

bbradelTT commented on tests/ttnn/unit_tests/gtests/ccl/test_1d_fabric_loopback_latency.cpp:
```diff
@@ -0,0 +1,488 @@
+        return;
+    }
+    if (arch == tt::ARCH::GRAYSKULL) {
+        log_info("Test must be run on WH");
```
I thought GS is not supported any more. Also, what about BH?

SeanNijjar replied:
resolved

bbradelTT commented on tests/tt_metal/microbenchmarks/ethernet/test_1d_fabric_latency.py:
```diff
@@ -0,0 +1,216 @@
+        },
+    }
+    devices_data = import_log_run_stats(setup)
+    print(f"{list(devices_data['devices'].keys())}")
```
Would it be better to use logger?

SeanNijjar replied:
resolved

SeanNijjar committed:
Update to add fused write atomic test

Also tighten up test to remove the extra atomic inc packet when we are sending packets with payloads larger than 0 B.

SeanNijjar committed:
missed an update

SeanNijjar committed:
middle of range

SeanNijjar committed:
relax max latency threshold

SeanNijjar committed:
the bounds finally seem stable enough

tt-rkim reviewed (approved):
will approve to unblock but please check if the superset data is being uploaded properly by checking the data artifacts at the bottom of the workflow run summary page so that we capture this test data

tt-snijjar committed:
split BW and latency tests into two jobs

williamlyTT commented:
I have a fix [here](https://github.com/tenstorrent/tt-metal/pull/20571) for the microbenchmark csv artifact.
@tt-rkim we're not uploading this pipeline to superset atm, should we turn that on? If so I can make the required workflow changes outside of this PR.
